rfcs/rfc-001-smart-contracts-engine.md

```diff
@@ -37,154 +37,175 @@ version 0.1
-
-* `Instance_AllocateContext(Instance PubKey, My PubKey, Proof of key ownership)`
-
-Any transaction can then call `call` on the contract with a set of keys.  It's up to the contract itself to manage ownership:
-
-* `Instance_Call(Instance PubKey, [Context PubKeys], proofs of ownership, userdata...)`
+At it's core, this is just a set of PublicKeys and Signatures with a bit of metadata.  The contract PublicKey routes this transaction into that contracts entry point.  `version` is used for dropping retransmitted requests.
 
 Contracts should be able to read any state that is part of solana, but only write to state that the contract allocated.
 
-#### Caller State
-
-Caller `state` is memory allocated for the `call` that belongs to the public key that is issuing the `call`.  This is the caller's context.
-
-#### Instance State
-
-Instance `state` is memory that belongs to this contract instance.  We may also need module-wide `state` as well.
-
-#### Participant State
+### Execution
 
-Participant `state` is any other memory.  In some cases it may make sense to have these allocated as part of the call by the caller.
+Calls batched and processed in a pipeline
 
-### Reduce
+```
++-----------+    +-------------+    +--------------+    +--------------------+    
+| sigverify |--->| lock memory |--->| validate fee |--->| allocate new pages |--->
++-----------+    +-------------+    +--------------+    +--------------------+    
+                                
+    +------------+    +---------+    +--------------+    +-=------------+   
+--->| load pages |--->| execute |--->|unlock memory |--->| commit pages |   
+    +------------+    +---------+    +--------------+    +--------------+   
 
-Some operations on the contract will require iteration over all the keys.  To make this parallelizable the iteration is broken up into reduce calls which are combined.
+```
 
+At the `execute` stage, the loaded pages have no data dependencies, so all the contracts can be executed in parallel. 
+## Memory Management
 ```
-void reduce_m(
-    const struct instance_data *data,
-    const uint8_t *keys[],
-    const uint8_t *data[],
-    int num,
-    uint8_t *reduce_data,
-);
-
-void reduce_r(
-    const struct instance_data *data,
-    const uint8_t *reduce_data[],
-    int num,
-    uint8_t *reduce_data,
-);
+pub struct Page {
+    /// key that indexes this page
+    /// proove ownership of this key to spend from this Page
+    owner: PublicKey,
+    /// contract that owns this page
+    /// contract can write to the data that is pointed to by `pointer`
+    contract: PublicKey,
+    /// balance that belongs to owner
+    balance: u64,
+    /// version of the structure, public for testing
+    version: u64,
+    /// hash of the page data
+    memhash: Hash,
+    /// The following could be in a separate structure
+    memory: Vec<u8>,
+}
 ```
 
-### Execution
+The guarantee that solana enforces:
```

Comment:
I'd prefer the docs not using branding terminology. In this case, replace "solana" with "the runtime".